m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+     */
+    public static class Builder {
+        private String[] hostPorts;
+        private String service;
+        private String env;
+        // Non-generic EMPTY ImmutableMap will never contain any elements
+        @SuppressWarnings("unchecked")
+        private ImmutableMap<String, String> commonTags = ImmutableMap.EMPTY;
+        private boolean includeHost;
+        private ThriftProtocol protocol = ThriftProtocol.COMPACT;
+        private int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE;
+        private int maxPacketSizeBytes = DEFAULT_MAX_PACKET_SIZE;
+        private String histogramBucketIdName = DEFAULT_HISTOGRAM_BUCKET_ID_NAME;
+        private String histogramBucketName = DEFAULT_HISTOGRAM_BUCKET_NAME;
+        private int histogramBucketTagPrecision = DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION;
+
+        private ExecutorService executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
+
+        /**
+         * Constructs a {@link Builder}. Having at least one host/port is required.
+         * @param hostPorts the array of host/ports
+         */
+        public Builder(String[] hostPorts) {
+            if (hostPorts == null || hostPorts.length == 0) {
+                throw new IllegalArgumentException("Must specify at least one host port");
+            }
+
+            this.hostPorts = hostPorts;
+        }
+
+        /**
+         * Configures the service of this {@link Builder}.
+         * @param service the value to set
+         * @return this {@link Builder} with the new value set
+         */
+        public Builder service(String service) {
+            this.service = service;
+
+            return this;
+        }
+
+        /**
+         * Configures the env of this {@link Builder}.
+         * @param env the value to set
+         * @return this {@link Builder} with the new value set
+         */
+        public Builder env(String env) {
+            this.env = env;
+
+            return this;
+        }
+
+        /**
+         * Configures the common tags of this {@link Builder}.
+         * @param commonTags the value to set
+         * @return this {@link Builder} with the new value set
+         */
+        public Builder commonTags(ImmutableMap<String, String> commonTags) {
+            this.commonTags = commonTags;
+
+            return this;
+        }
+
+        /**
+         * Configures whether to include the host tag of this {@link Builder}.
+         * @param includeHost the value to set
+         * @return this {@link Builder} with the new value set
+         */
+        public Builder includeHost(boolean includeHost) {
+            this.includeHost = includeHost;
+
+            return this;
+        }
+
+        /**
+         * Configures the protocol of this {@link Builder}.
+         * @param protocol the value to set
+         * @return this {@link Builder} with the new value set
+         */
+        public Builder protocol(ThriftProtocol protocol) {
```

Comment:
will users ever override this?

Reply:
Yes, they can specify `ThriftProtocol.BINARY` as opposed to the default `TBinaryProtocol.COMPACT`, which will make the reporter use that TProtocol accordingly.

Reply:
Yeah but will users ever have to set this to the non-default value? I'd rather not reveal the transport mechanism as a choice to the users if we can help it.

Reply:
It seems like TCompactProtocol is the more efficient protocol here, so I'll just default to using that.